subsys/fs/nvs/nvs.c

```diff
@@ -0,0 +1,803 @@
+{
+	struct _nvs_sector_hdr sector_hdr;
+	int rc;
+
+	/* just to be sure, align offset to sector_size */
+	offset &= ~(fs->sector_size - 1);
+	fs->sector_id += 1;
+	rc = _nvs_sector_hdr_get(fs, offset, &sector_hdr);
+	if (rc) {
+		return NVS_ERR_FLASH;
+	}
+	if (sector_hdr.fd_magic != 0xFFFFFFFF) {
+		return NVS_ERR_NOSPACE;
+	}
+	sector_hdr.fd_magic = fs->magic;
+	sector_hdr.fd_id = fs->sector_id;
+	sector_hdr._pad = 0;
+	rc = nvs_flash_write(fs,
+		offset,
+		&sector_hdr,
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr)));
+	if (rc) {
+		return NVS_ERR_FLASH;
+	}
+	fs->write_location = offset +
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
+	return NVS_OK;
+}
+
+/* check if a sector is used by looking at all elements of the sector.
+ * return 0 if sector is empty, 1 if it is not empty
+ */
+int _nvs_sector_is_used(struct nvs_fs *fs, off_t offset)
+{
+	int rc;
+	off_t addr;
+	u8_t buf;
+
+	rc = 0;
+	offset &= ~(fs->sector_size - 1);
+	for (addr = 0; addr < fs->sector_size; addr += sizeof(buf)) {
+		rc = flash_read(fs->flash_device, fs->offset + addr,
+			&buf, sizeof(buf));
+		if (rc) {
+			return NVS_ERR_FLASH;
+		}
+		if (buf != 0xFF) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+/* erase the flash starting at offset, erase length = len */
+int _nvs_flash_erase(struct nvs_fs *fs, off_t offset, size_t len)
+{
+	int rc;
+
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return NVS_ERR_FLASH;
+	}
+
+	rc = flash_erase(fs->flash_device, fs->offset + offset, len);
+	if (rc) {
+		/* flash write error */
+		return NVS_ERR_FLASH;
+	}
+
+	DBG_NVS("Erasing flash at %x, len %x\n", offset, len);
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	return NVS_OK;
+}
+
+void _nvs_entry_sector_advance(struct nvs_fs *fs)
+{
+	fs->entry_sector++;
+	if (fs->entry_sector == fs->sector_count) {
+		fs->entry_sector = 0;
+	}
+}
+
+/* garbage collection: addr is set to the start of the sector to be gc'ed,
+ * the entry sector has been updated to point to the sector just after the
+ * sector being gc'ed
+ */
+int _nvs_gc(struct nvs_fs *fs, off_t addr)
+{
+	int rc, len, bytes_to_copy;
+	off_t rd_addr;
+	struct nvs_entry walker, walker_last, last_entry, search;
+	struct _nvs_data_hdr head;
+	u8_t buf[NVS_MOVE_BLOCK_SIZE];
+
+	walker.data_addr = addr;
+	_nvs_addr_advance(fs,
+		&walker.data_addr,
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr)) +
+		_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+	while (1) {
+		rc = nvs_flash_read(fs,
+			_nvs_head_addr_in_flash(fs, &walker),
+			&head,
+			_nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr)));
+		if (rc) {
+			return NVS_ERR_FLASH;
+		}
+		if (head.id == NVS_ID_EMPTY) {
+			return NVS_ERR_NOVAR;
+		}
+		if (head.id == NVS_ID_SECTOR_END) {
+			return NVS_OK;
+		}
+		walker.len = head.len;
+		walker.id = head.id;
+		search.id = walker.id;
+		if (nvs_get_first_entry(fs, &search) != NVS_OK) {
```

Comment:
Dose it scan over all NVS flash sectors? Shouldn't it be limited to current sector?

Reply:
got it. It's ok here.

Reply:
I was just preparing an answer: it scans over all sectors except the one that is prepared for deletion.